Integration of kernels and extensions with jupyter

costrouc opened the pull request:
###### Motivation for this change

**this main comment has been edited to reflect progress**

Anyone who has used `jupyterlab` and `notebooks` has run into the issue of adding kernels and  extensions to the UI. This is tricky because often times the kernels run in a different language e.g. `julia`, `haskell`, etc. Nix can handle this extremely well and would make jupyter notebook installations much easier. This PR provides intructions on a way to "auto-install` kernels and extensions into the nix environment. Any input would be much appreciated. Here is what my research has shown me so far (I dont understand extensions yet where `static` directory is put).

# Jupyter Kernels

The environment variable `JUPYTER_PATH` is a list of `:` separated paths that have a `kernels` directory. Every directory within contains kernels `kernels/<kernel-name>/kernel.json`. Here is an example `kernel.json`.

```json
{
    "argv":["/nix/store/bwfygfcdvis9wd1c1v51xwnwhw1hx0a0-python3-3.6.6/bin/python3.6m","-m","jupyter_c_kernel","-f","{connection_file}"],
    "display_name":"C",
    "language":"c"
}
```

I would like to automatically add a set of kernels via the `JUPYTER_PATH` variable to the `jupyterlab` and `notebook` packages. Not sure exactly how to do this here is an example of me setting the variable after the fact in a `nix-shell`. It has been show below how to automatically do this but the issue is that kernels exist for many languages not just python. What I have settled on is that `kernel.json` and the images should be included at `share/jupyter/kernels/<kernel-name>/`.

# JupyterLab Extensions

Jupyterlab is the future so I have put the most effort into installation instructions. Additionally `notebook`'s installation paths are not as well documented and often times work without any additional steps. Jupyterlab has a command line option `--app-dir` and environment variable `JUPYTERLAB_DIR` that control where the extensions are installed usually `share/jupyter/lab`. These extensions are `npm` packages that after being combined through a `package.json` must be "compiled" with webpack to produce compressed static files. The npm package dependencies can easily be in the thousands... thanks Javascript leading to about 100MB - 1GB of dependencies that must be compiled for each additional extension added. This step cannot be pure unless we override the jupyterlab package management and handle all of the `npm` dependencies. Currently it seems that `npm` packages are not a good fit for `nixpkgs`. So I have come up with an `unpure` but working way to package jupyterlab extensions. Each package that comes with a jupyterlab extension uses `passthru = { jupyterlabExtensions = [ .... ]; };`. Then we include the following lines 

```nix
# make temp directory for extensions
TEMPDIR=$(mktemp -d -p /tmp)
mkdir -p $TEMPDIR
# copy base jupyter lab setup
cp -r ${pkgs.python36Packages.jupyterlab}/share/jupyter/lab/* $TEMPDIR
chmod -R 755 $TEMPDIR

# gather unique jupyterlab extensions
# for each extension
#      jupyter labextension install --no-build --app-dir=$TEMPDIR <extension>
${pkgs.stdenv.lib.concatMapStrings
     (s: "jupyter labextension install --no-build --app-dir=$TEMPDIR ${s}; ")
     (pkgs.lib.unique
       ((pkgs.lib.concatMap
           (d: pkgs.lib.attrByPath ["passthru" "jupyterlabExtensions"] [] d)
           buildInputs) ++ additionalExtensions))  }
# finally build jupyterlab npm extensions with webpack
jupyter lab build --app-dir=$TEMPDIR

# start jupyterlab service
jupyter lab --app-dir=$TEMPDIR
```

# Working Example with 20 jupyerlab extension + 4 kernels (c, python, go, ansible)

All that needs to be edited by user is `kernels`, `additionalExtensions`, and `buildInputs`. The rest is automatic and will launch a jupyterlab instance for you.

`shell.nix`
```nix
{ pkgs ? import <nixpkgs> {}, pythonPackages ? pkgs.python36Packages }:

let kernels = [
      pkgs.python36Packages.ansible-kernel
      pythonPackages.jupyter-c-kernel
      pkgs.gophernotes
    ];

    additionalExtensions = [
      "@jupyterlab/toc"
      "@jupyterlab/fasta-extension"
      "@jupyterlab/geojson-extension"
      "@jupyterlab/katex-extension"
      "@jupyterlab/mathjax3-extension"
      "@jupyterlab/plotly-extension"
      "@jupyterlab/vega2-extension"
      "@jupyterlab/vega3-extension"
      "@jupyterlab/xkcd-extension"
      "jupyterlab-drawio"
      "@jupyterlab/hub-extension"
      "jupyterlab_bokeh"
    ];
in
pkgs.mkShell rec {
  buildInputs = [
    ### Base Packages
    pythonPackages.jupyterlab pkgs.nodejs

    ### Extensions
    pythonPackages.ipywidgets
    pythonPackages.ipydatawidgets
    pythonPackages.ipywebrtc
    pythonPackages.pythreejs
    pythonPackages.ipyvolume
    pythonPackages.jupyterlab-git
    pythonPackages.jupyterlab-latex
    pythonPackages.ipyleaflet
    pythonPackages.ipympl
  ] ++ kernels;

  shellHook = ''
    TEMPDIR=$(mktemp -d -p /tmp)
    mkdir -p $TEMPDIR
    cp -r ${pkgs.python36Packages.jupyterlab}/share/jupyter/lab/* $TEMPDIR
    chmod -R 755 $TEMPDIR
    echo "$TEMPDIR is the app directory"

    # kernels
    export JUPYTER_PATH="${pkgs.lib.concatMapStringsSep ":" (p: "${p}/share/jupyter/") kernels}"

# labextensions
${pkgs.stdenv.lib.concatMapStrings
     (s: "jupyter labextension install --no-build --app-dir=$TEMPDIR ${s}; ")
     (pkgs.lib.unique
       ((pkgs.lib.concatMap
           (d: pkgs.lib.attrByPath ["passthru" "jupyterlabExtensions"] [] d)
           buildInputs) ++ additionalExtensions))  }
jupyter lab build --app-dir=$TEMPDIR

# start jupyterlab
jupyter lab --app-dir=$TEMPDIR
  '';

}
```

###### Things done

Not all of these kernels work currently (gnuplot, zeus-cling)

Adding four kernels:
  - [X] [jupyter_c_kernel](https://github.com/brendan-rius/jupyter-c-kernel) (C)
  - [X] [ansible_kernel](https://github.com/ansible/ansible-jupyter-kernel) (ansible)
  - [X] [gophernotes](https://github.com/gopherdata/gophernotes) (Go)

Want to add extension support for python/jupyterlab extension:
  - ipywebrtc (webrtc in jupyterlab)
  - pythreejs (three.js in jupyterlab)
  - ipyvolume (3d volumetric rendering in jupyterlab( 
  - jupyterlab-latex (latex document editing in jupyterlab)
  - ipywidgets (widgets alot of packages depend on this one)
  - jupyterlab-git (git support)
  - ipyleaflet (leaflet extension)
  - ipympl (matplotlib extension)
  - ipydatawidgets (A set of widgets to help facilitate reuse of large datasets)

This will require input on how to properly include extensions and kernels within `jupyter` derivations.

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

costrouc committed:
pythonPackages.ansible-kernel: init at 0.8.0

costrouc committed:
pythonPackages.ansible-runner: init at 1.1.2

costrouc committed:
pythonPackages.jupyter-c-kernel: init at 1.2.2

FRidh commented:
You have seen the top-level `jupyter` and `jupyter-kernel` attributes that are used by the `jupyter` NixOS module?

costrouc commented:
I had not seen that before and thank you for pointing that out! However, I see some issues with declaring this in `nixos` modules in that only `nixos` can use that. Ideally I would like to say "start a jupyter notebook with these 3 kernels (julia, C, and python) and these 3 extensions (ipywidget, ipyvolume, jupyterlab-drawio)" for example. This currently is not possible and a pain point I have had with jupyter. Welcome to any ideas. 

I have learned that the `JUPYTER_PATH` environment variable can point to each kenel `json` definition and `JUPYTERLAB_DIR` points to the javascript extensions static files. This will require a meta package or something to collect all of the extensions in one location and set `JUPYTER_PATH` correctly.

FRidh commented:
> I had not seen that before and thank you for pointing that out! However, I see some issues with declaring this in nixos modules in that only nixos can use that. Ideally I would like to say "start a jupyter notebook with these 3 kernels (julia, C, and python) and these 3 extensions (ipywidget, ipyvolume, jupyterlab-drawio)" for example in any linux distribution. This currently is not possible and a pain point I have had with jupyter. Welcome to any ideas.

The NixOS module only provides an interface, the actual logic is implemented in the two top-level attributes I referred to.

FRidh commented:
Original PR https://github.com/NixOS/nixpkgs/pull/33673

costrouc commented:
Thank you and I should read more carefully the expressions. I will try it out with the new kernels with the derivation and also would be interested in adding extensions support to the `nixpkgs/pkgs/applications/editors/jupyter/default.nix`.

FRidh commented on pkgs/development/python-modules/gnuplot-kernel/default.nix:
```diff
@@ -0,0 +1,50 @@
+  };
+
+  checkInputs = [ pytest ];
+  propagatedBuildInputs = [ notebook metakernel gnuplot ];
```
Should we not somewhere hardcode a path to gnuplot?

costrouc replied:
Yes we should. Currently I am trying to get the `gnuplot_kernel` to work. It freezes on me for an unknown reason.

veprbl commented:
> I would like to automatically add a set of kernels via the `JUPYTER_PATH` variable to the `jupyterlab` and `notebook` packages.

This should be easy to achieve by adding a `setupHook` to `jupyterlab`.
Examples are

 * cmake: https://github.com/NixOS/nixpkgs/blob/a6809c0dd0694ac0c1268ef6420f5c81edcc406a/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L1-L3
https://github.com/NixOS/nixpkgs/blob/a6809c0dd0694ac0c1268ef6420f5c81edcc406a/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L94
 * python:
https://github.com/NixOS/nixpkgs/blob/a6809c0dd0694ac0c1268ef6420f5c81edcc406a/pkgs/development/interpreters/python/setup-hook.sh#L1-L3
https://github.com/NixOS/nixpkgs/blob/a6809c0dd0694ac0c1268ef6420f5c81edcc406a/pkgs/development/interpreters/python/setup-hook.sh#L15

costrouc committed:
gophernotes: init at 1.0.0-RC1

costrouc committed:
pythonPackages.nbval: init at 0.9.1

costrouc committed:
pythonPackages.nbdime: init at 1.0.3

costrouc committed:
pythonPackages.ipywebrtc: init at 0.4.1

costrouc committed:
pythonPackages.pythreejs: init at 2.0.2

costrouc committed:
pythonPAckages.traittypes: init at 0.2.1

costrouc committed:
pythonPackages.pytest-tornado5: init at 1.0.0

costrouc committed:
pythonPackages.ipydatawidgets: init at 4.0.0

costrouc committed:
pythonPackages.ipyvolume: init at 0.5.1

costrouc commented:
`ansible-kernel` depends on https://github.com/NixOS/nixpkgs/pull/49764 being move to `python-modules` for python 2.7 and 3+ support. 

Here is the error I get (due to using python2.7 for 3 packages)

```
  File "/nix/store/bwfygfcdvis9wd1c1v51xwnwhw1hx0a0-python3-3.6.6/lib/python3.6/re.py", line 142, in <module>
    class RegexFlag(enum.IntFlag):
AttributeError: module 'enum' has no attribute 'IntFlag'
```

costrouc committed:
pythonPackages.jupytext: init at 0.8.4

costrouc committed:
pythonPackages.asteval: init at 0.9.13

costrouc committed:
pythonPackages.lmfit: init at 0.9.11

costrouc committed:
pythonPackages.jupyterlab-git: init at 0.3.0

costrouc committed:
pythonPackages.jupyterlab-latex: init at 0.4.1

costrouc committed:
pythonPackages.ipywidgets: refactor add jupyterlab extension

costrouc committed:
pythonPackages.ipyleaflet: init at 0.9.1

costrouc committed:
pythonPackages.ipympl: init at 0.2.1

costrouc commented:
Still a work in progress but I have support for jupyterlab extensions. Though it is not "pure". This is because the extensions rely heavily of npm and require a webpack rebuild. Here is a working shell with three kernels and about 20 extensions.

```
{ pkgs ? import <nixpkgs> {}, pythonPackages ? pkgs.python36Packages }:

let kernels = [
      # # requires PR on ansible to be moved to python-modules
      # pkgs.python36Packages.ansible-kernel
      pythonPackages.jupyter-c-kernel
      pkgs.gophernotes
      # pkgs.ihaskell
    ];

    additionalExtensions = [
      "@jupyterlab/toc"
      # jupyter-renderers
      "@jupyterlab/fasta-extension"
      "@jupyterlab/geojson-extension"
      "@jupyterlab/katex-extension"
      "@jupyterlab/mathjax3-extension"
      "@jupyterlab/plotly-extension"
      "@jupyterlab/vega2-extension"
      "@jupyterlab/vega3-extension"
      "@jupyterlab/xkcd-extension"
      "jupyterlab-drawio"
      "@jupyterlab/hub-extension"
      "jupyterlab_bokeh"
    ];
in
pkgs.mkShell rec {
  buildInputs = [
    ### Base Packages
    pythonPackages.jupyterlab pkgs.nodejs

    ### Extensions
    pythonPackages.ipywidgets
    pythonPackages.ipydatawidgets
    pythonPackages.ipywebrtc
    pythonPackages.pythreejs
    pythonPackages.ipyvolume
    pythonPackages.jupyterlab-git
    pythonPackages.jupyterlab-latex
    pythonPackages.ipyleaflet
    pythonPackages.ipympl
  ] ++ kernels;

  shellHook = ''
    TEMPDIR=$(mktemp -d -p /tmp)
    mkdir -p $TEMPDIR
    cp -r ${pkgs.python36Packages.jupyterlab}/share/jupyter/lab/* $TEMPDIR
    chmod -R 755 $TEMPDIR
    echo "$TEMPDIR is the app directory"

    # kernels
    export JUPYTER_PATH="${pkgs.lib.concatMapStringsSep ":" (p: "${p}/share/jupyter/") kernels}"

    # labextensions
    ${pkgs.stdenv.lib.concatMapStrings
         (s: "jupyter labextension install --no-build --app-dir=$TEMPDIR ${s}; ")
         (pkgs.lib.unique
            (pkgs.lib.concatMap
               (d: pkgs.lib.attrByPath ["passthru" "jupyterlabExtensions"] [] d)
               buildInputs)) ++ additionalExtensions }
    jupyter lab build --app-dir=$TEMPDIR

    # start jupyterlab
    jupyter lab --app-dir=$TEMPDIR
  '';

}
```

costrouc commented:
@veprbl I am not very knowledgeable on how to properly do this. Could you give some further advice? Additionally I should state that jupyterlab kernels are not necessarily for python only. For example this PR adds support for `Go` and `C`.

costrouc commented:
I would like some feedback if

```nix
...
passthru ={
   jupyterlabExtensions = [ "jupyterlab-datawidgets" "@jupyter-widgets/jupyterlab-manager
" ];
};
...
```

Is the correct way to pass that the package requires an `npm` dependency that cannot be included in the build requirements. Is this a good way to pass these dependencies? This expression is then used to select the necessary extensions to install (`additionallExtensions` is a list of strings of extensions that do not require a python package).

```nix
(pkgs.lib.unique
       ((pkgs.lib.concatMap
           (d: pkgs.lib.attrByPath ["passthru" "jupyterlabExtensions"] [] d)
           buildInputs) ++ additionalExtensions))
```

FRidh commented:
> I would like some feedback of using passthru

Looks good to me.

costrouc commented:
With that I would say that this is ready for merge. It does not change any packages only adding new ones.

Also I will be adding documentation on the nixos python wiki on using extensions and kernels with jupyterlab. Wanted to document it for others.

veprbl commented:
> @veprbl I am not very knowledgeable on how to properly do this. Could you give some further advice?

I can only mention that there is some info on hooks in the nixpkgs manual https://nixos.org/nixpkgs/manual/#ssec-setup-hooks . If you have any specific problems with using `setupHook` feel free to create an issue.

Having said that, I think it would be fine to limit the scope of this PR to just adding the new packages. Your solution with setting `JUPYTER_PATH` in the `shellHook` should work, though it does require to have a `shell.nix`.

costrouc commented:
@veprbl agreed. In a future PR I would certainly like to look at automating with a setup-hook. I appreciate you sharing the documentation.

costrouc commented:
This PR is ready for merge. Was checked with `nix-review`. Only `python36Packges.ansible-kernel` does not build. This package will build once `ansible` is given python 3 support in https://github.com/NixOS/nixpkgs/pull/49764.

GrahamcOfBorg commented:
<!--REQUEST_ID=37868768-678c-403e-9a2f-9a4db235fc85-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49807&attempt_id=3e5e0c54-2fed-4e8d-b45e-fcc5471932fe)

Attempted: gophernotes

<details><summary>Partial log (click to expand)</summary><p>

```
}
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/lpx96i4sm83xb5v1admld75zj7llyahp-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
shrinking /nix/store/lpx96i4sm83xb5v1admld75zj7llyahp-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin/bin/gophernotes
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/lpx96i4sm83xb5v1admld75zj7llyahp-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin/bin
patching script interpreter paths in /nix/store/lpx96i4sm83xb5v1admld75zj7llyahp-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
checking for references to /build in /nix/store/lpx96i4sm83xb5v1admld75zj7llyahp-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin...
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
/nix/store/lpx96i4sm83xb5v1admld75zj7llyahp-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=37868768-678c-403e-9a2f-9a4db235fc85-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49807&attempt_id=fe3e4599-d049-4914-9431-0c65da552e4a)

Attempted: gophernotes

<details><summary>Partial log (click to expand)</summary><p>

```
}
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/dynzgchgqfii1c0h6vqk6v0jxcwcxsgg-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
shrinking /nix/store/dynzgchgqfii1c0h6vqk6v0jxcwcxsgg-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin/bin/gophernotes
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/dynzgchgqfii1c0h6vqk6v0jxcwcxsgg-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin/bin
patching script interpreter paths in /nix/store/dynzgchgqfii1c0h6vqk6v0jxcwcxsgg-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
checking for references to /build in /nix/store/dynzgchgqfii1c0h6vqk6v0jxcwcxsgg-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin...
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
/nix/store/dynzgchgqfii1c0h6vqk6v0jxcwcxsgg-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
```
</p></details>

MMesch commented:
@costrouc the PR looks really nice! Thanks for you effort.

I have a question regarding this part:

> These extensions are npm packages that after being combined through a package.json must be "compiled" with webpack to produce compressed static files. The npm package dependencies can easily be in the thousands... thanks Javascript leading to about 100MB - 1GB of dependencies that must be compiled for each additional extension added. This step cannot be pure unless we override the jupyterlab package management and handle all of the npm dependencies.

If I understand correctly, jupyterlab is bundled with `jupyter lab build` into one static javascript application that includes all internal node modules and all external extensions. This static jupyterlab bundle is placed somewhere in the current JUPYTERLAB_DIR. One example where jupyterlab is built together with additional extensions outside of its base folder is described here https://github.com/jupyterlab/jupyterlab/issues/4064.

It would be nice to have this as a single `jupyterlabWith` prebundled derivation that can be loaded instantly in a nix shell. Do you see any way to achieving this. Would node2nix help achieving this? Can we define a single `node-packages.nix` that includes all jupyterlab extensions? How did you get to the 100Mb-1Gb estimate?

costrouc commented:
> If I understand correctly, jupyterlab is bundled with jupyter lab build into one static javascript application that includes all internal node modules and all external extensions. This static jupyterlab bundle is placed somewhere in the current JUPYTERLAB_DIR. 

This is pretty much correct. `JUPYTERLAB_DIR` and the command line option `--appdir` have the same function. Within this directory are several directories. The important ones here are `staging` and `static`. `staging` is the location where all of the `npm` packages are installed into a `node_modules` folder. The staging directory is used for building only and gets quite large. The 100 MB - 1 GB is what I got from a minimal install with `ipywidgets` and  the latter is with everything including the kitchen sink installed. `static` is for the compiled `jupyter labextension build` which is never larger than 50 Mb in my experience.

> It would be nice to have this as a single jupyterlabWith prebundled derivation that can be loaded instantly in a nix shell. Do you see any way to achieving this. Would node2nix help achieving this? Can we define a single node-packages.nix that includes all jupyterlab extensions?

This 100%. I would like it too but the blocker is `npm`. For one I am not the most experienced with it so it may have the feature that I am looking for. The issue is that in jupyterlab you want to choose a subset of extensions to install from all extensions. This means that somehow `node2nix` needs to support choosing several packages to include. Did that make any sense? I am not sure I explained it correctly. If that features exists and we added the `npm` packages to `nixpkgs` jupyterlabWith is certainly possible and is what I would want to do.

Are you experienced with `node2nix` and would it support such a capability?

costrouc commented:
I was unaware of `link` and how it works but if it were able to isolate each extension this that might just solve it. Looks promising. I would really appreciate help in this area but I will be looking into it as well. Anything we can do to make the extensions a derivation is perfect

costrouc committed:
adding jupterlab extensions

Extensions:
 - ipyvolume
 - @jupyter-widgets/jupyterlab-manager
 - @jupyterlab/fasta-extension
 - @jupyterlab/geojson-extension
 - @jupyterlab/git
 - @jupyterlab/hub-extension
 - @jupyterlab/katex-extension
 - @jupyterlab/latex
 - @jupyterlab/mathjax3-extension
 - @jupyterlab/plotly-extension
 - @jupyterlab/toc
 - @jupyterlab/vega2-extension
 - @jupyterlab/vega3-extension
 - @jupyterlab/xkcd-extension
 - jupyter-leaflet
 - jupyter-matplotlib
 - jupyter-threejs
 - jupyter-webrtc
 - jupyterlab-datawidgets
 - jupyterlab-drawio

MMesch commented:
agree. I'll help you with the PR ...

GrahamcOfBorg commented:
<!--REQUEST_ID=60c5ad87-cc92-4987-83e4-ae5ee6fbb12c-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49807&attempt_id=8f1dfb5d-66dc-4d73-bcdb-a89d3c12d6b8)

Attempted: gophernotes

<details><summary>Partial log (click to expand)</summary><p>

```
}
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/lwkcg6bzksgpl9fnqx7sw97b8s1kblk8-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
shrinking /nix/store/lwkcg6bzksgpl9fnqx7sw97b8s1kblk8-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin/bin/gophernotes
strip is /nix/store/rpbg8gmqxhz8g61p1plz5d2srs84pvmv-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/lwkcg6bzksgpl9fnqx7sw97b8s1kblk8-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin/bin
patching script interpreter paths in /nix/store/lwkcg6bzksgpl9fnqx7sw97b8s1kblk8-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
checking for references to /build in /nix/store/lwkcg6bzksgpl9fnqx7sw97b8s1kblk8-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin...
strip is /nix/store/rpbg8gmqxhz8g61p1plz5d2srs84pvmv-binutils-2.30/bin/strip
/nix/store/lwkcg6bzksgpl9fnqx7sw97b8s1kblk8-gophernotes-unstable-b58dd906c43c1ef0b5976b150872c79f82daafcd-bin
```
</p></details>

costrouc commented:
I will also be closing this along with two other PRs to combine with this one. https://github.com/NixOS/nixpkgs/pull/49807,  https://github.com/NixOS/nixpkgs/pull/49721, and https://github.com/NixOS/nixpkgs/pull/50858.

This will lead to a PR with better python 3 support for jupyter (newest releases are python 3.5+ only). Addition of 3 kernels (c, go, ansible) and 20 jupyterlab extensions. Also will have an editor that has a clean way to add kernels and extensions. This will be a big PR but I am finding that they all depend on each other.

ariutta commented:
> This will lead to a PR...

Did the new PR land?

Related: it appears similar functionality can be found in the [JupyterWith repo](https://github.com/tweag/jupyterWith).
